[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147824

Earlier, with inline flag we were lifting id-guarded tensors to the inputs to the Fx graph. But this offers no benefit. Main idea behind lifting parameters as inputs was to reuse the compilation units across many instances of the nn-module. However, if we are guarding on the `id`, we are explicitly specializing the compiled artifact to the parameter.

This PR installs the parameters back into the graph. The benefit is removal of all pre-graph bytecode to extract the id-guarded tensors from locals/globals. This increases speedup from 1.67x to 1.75x for an internal model that has large number of optimizer parameters.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

anijain2305 committed:
[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/optimizer.py:
```diff
@@ -358,7 +358,7 @@ def wrap_tensor(self, tx: "InstructionTranslator", tensor_value):
             mark_static_address(tensor_value)
             source = self.tensor_to_source[tensor_value]
-            self.static_tensor_names.add(tx.output.module_key_name(source.name))
+            self.static_tensor_names.add(tx.output.module_key_name(source.name()))
```
pre-existing bug.

anijain2305 committed:
Update on "[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jansel commented on torch/_dynamo/output_graph.py:
```diff
@@ -1556,6 +1563,11 @@ def example_inputs(self) -> list[torch.Tensor]:
         return result
 
+    def remove_unused_get_attr_nodes(self) -> None:
+        for node in reversed(list(self.graph.nodes)):
+            if node.op == "get_attr" and len(list(node.users)) == 0:
```
```suggestion
        for node in sorted(self.graph.find_nodes(op="get_attr"), reverse=True):
            if len(list(node.users)) == 0:
```
So we don't need to take a pass over the entire graph.

jansel replied:
Not fixed?

anijain2305 committed:
Update torch/_dynamo/output_graph.py

Co-authored-by: Jason Ansel <jansel@meta.com>

anijain2305 commented:
@pytorchbot rebase

anijain2305 committed:
Update

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][optimizers] Install ID_GUARDED tensors into the Fx graph"


Earlier, with inline flag we were lifting id-guarded tensors to the inputs to the Fx graph. But this offers no benefit. Main idea behind lifting parameters as inputs was to reuse the compilation units across many instances of the nn-module. However, if we are guarding on the `id`, we are explicitly specializing the compiled artifact to the parameter.

This PR installs the parameters back into the graph. The benefit is removal of all pre-graph bytecode to extract the id-guarded tensors from locals/globals. This increases speedup from 1.67x to 1.75x for an internal model that has large number of optimizer parameters.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge -f

anijain2305 commented:
@pytorchbot merge -f "stuck CI"

anijain2305 commented:
This could be related to https://github.com/pytorch/pytorch/issues/152275
